### PR TITLE
[SGD] Add PTL Docs

### PR DESCRIPTION
**Pull request by amogkam**
Long overdue LightningOperator documentation

image
image
image
image
image
image

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/requirements-doc.txt [hunk 1]
```diff
@@ -13,6 +13,7 @@ pickle5
 pillow
 pydantic<1.7
 pygments
+pytorch-lightning
```
Comment (richardliaw) [inline doc/requirements-doc.txt:16]: can we avoid making pytorch lightning a requirement of the documentation? we have tried to keep this build as minimal as possible

File: doc/source/index.rst [hunk 2]
```diff
@@ -274,6 +274,7 @@ Papers
    raysgd/raysgd_pytorch.rst
    raysgd/raysgd_tensorflow.rst
    raysgd/raysgd_dataset.rst
+   raysgd/raysgd_ptl.rst
```
Comment (richardliaw) [inline doc/source/index.rst:277]: fine for now, but I think we will want to refactor the raysgd documentation soon

File: doc/source/raysgd/raysgd_ref.rst [hunk 3]
```diff
@@ -1,51 +1,80 @@
...
+CreatorOperator
+~~~~~~~~~~~~~~~~
+
+.. autoclass:: ray.util.sgd.torch.training_operator.CreatorOperator
+    :members:
+    :exclude-members: setup
+
+.. _ref-lightning-operator:
+
+Pytorch LightningOperator
```
Comment (richardliaw) [inline doc/source/raysgd/raysgd_ref.rst:34]:
```suggestion
PytorchLightning Operator
```

File: doc/source/raysgd/raysgd_ref.rst [hunk 4]
```diff
@@ -1,51 +1,80 @@
...
 Dataset
--------
+~~~~~~~
 
 .. autoclass:: ray.util.sgd.data.Dataset
     :members:
 
     .. automethod:: __init__
 
+Ray SGD Utils
```
Comment (richardliaw) [inline doc/source/raysgd/raysgd_ref.rst:72]: RaySGD one word -- this is intentional to make SEO easy

File: python/ray/util/sgd/torch/training_operator.py [hunk 5]
```diff
@@ -777,7 +777,15 @@ def from_ptl(cls,
                  lightning_module_cls,
                  train_dataloader=None,
                  val_dataloader=None):
-        """Creates a TrainingOperator from a Pytorch Lightning Module.
+        """A utility method to create a custom TrainingOperator class from a
+        Pytorch Lightning Module.
```
Comment (richardliaw) [inline python/ray/util/sgd/torch/training_operator.py:781]: pydoc 1 line summaries should only be 1 line

File: python/ray/util/sgd/torch/training_operator.py [hunk 6]
```diff
@@ -930,7 +945,20 @@ def scheduler_step_freq(self):
 
 class CreatorOperator(TrainingOperator):
     """A subclass of TrainingOperator specifically for defining training
-    state using creator functions.
+    state using creator functions. This allows for backwards compatibility
+    with pre Ray 1.0 versions.
```
Comment (richardliaw) [inline python/ray/util/sgd/torch/training_operator.py:949]: should be 1 line summary. You can break this into two paragraphs. (pydoc style)

File: doc/source/raysgd/raysgd_ptl.rst [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+.. _ptl-unsupported-features:
+
+Unsupported Features
+--------------------
+This integration is currently under active development, so not all Pytorch Lightning features are supported.
+Please post any feature requests on `Github
+<<https://github.com/ray-project/ray/issues>`_ and we will get to it shortly!
+
+A list of unsupported model hooks is as follows:
```
Comment (richardliaw) [inline doc/source/raysgd/raysgd_ptl.rst:113]: you should document when this list was updated.

File: doc/source/raysgd/raysgd_ptl.rst [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+    ray.init()
+    train_mnist(num_workers=32, use_gpu=True, num_epochs=5)
+
+.. _ptl-unsupported-features:
+
+Unsupported Features
+--------------------
+This integration is currently under active development, so not all Pytorch Lightning features are supported.
+Please post any feature requests on `Github
+<<https://github.com/ray-project/ray/issues>`_ and we will get to it shortly!
```
Comment (richardliaw) [inline doc/source/raysgd/raysgd_ptl.rst:111]:
```suggestion
<https://github.com/ray-project/ray/issues>`_ and we will get to it shortly!
```

File: doc/source/raysgd/raysgd_ptl.rst [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+.. code-block:: python
+
+    import ray
+    from ray.util.sgd.torch import TorchTrainer
+
+    ray.init()
+    trainer = TorchTrainer(training_operator_cls=MyLightningOperator, num_workers=4, use_gpu=True)
+    train_stats = trainer.train()
+
+And that's it! For a more comprehensive guide, see the MNIST tutorial below.
```
Comment (richardliaw) [inline doc/source/raysgd/raysgd_ptl.rst:48]: link ref to mnist tutorial

**Review by richardliaw — CHANGES_REQUESTED** [review]
Generally looks good, but we should avoid introducing pytorch-lightning as a dependency for the doc build.

Also, there were a couple errors when i tried building myself.

File: doc/requirements-doc.txt (hunk 1 above)
Comment (amogkam) [reply doc/requirements-doc.txt]: yes I removed it. It's not even needed.

File: python/ray/util/sgd/torch/training_operator.py (hunk 5 above)
Comment (amogkam) [reply python/ray/util/sgd/torch/training_operator.py]: done

File: python/ray/util/sgd/torch/training_operator.py (hunk 6 above)
Comment (amogkam) [reply python/ray/util/sgd/torch/training_operator.py]: done

File: doc/source/raysgd/raysgd_ptl.rst (hunk 7 above)
Comment (amogkam) [reply doc/source/raysgd/raysgd_ptl.rst]: done

File: doc/source/raysgd/raysgd_ptl.rst (hunk 9 above)
Comment (amogkam) [reply doc/source/raysgd/raysgd_ptl.rst]: I linked it. It's literally right below though.

**Comment by amogkam** [comment]
@richardliaw I addressed all the comments

**Review by richardliaw — APPROVED** [review]

**Comment by richardliaw** [comment]
nice work!

**Pull request merged**